<a href="https://colab.research.google.com/github/Ritwik510/Poem/blob/main/Project_Week_1_Solved%20(21016).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Load the necessary libraries. Import and load the dataset with a name uber_drives .

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import numpy as np
import pandas as pd
import plotly.express as px

dataset = pd.read_csv("/content/drive/My Drive/Value Added/Projects/Project Week1.csv")

## Q1. Show the last 10 records of the dataset. (2 point)

In [ ]:
dataset.tail(10)

START_DATE*         END_DATE* CATEGORY*            START*  \
1145  12/30/2016 10:15  12/30/2016 10:33  Business           Karachi   
1146  12/30/2016 11:31  12/30/2016 11:56  Business           Karachi   
1147  12/30/2016 15:41  12/30/2016 16:03  Business           Karachi   
1148  12/30/2016 16:45  12/30/2016 17:08  Business           Karachi   
1149  12/30/2016 23:06  12/30/2016 23:10  Business           Karachi   
1150   12/31/2016 1:07   12/31/2016 1:14  Business           Karachi   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Karachi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   

                 STOP*  MILES*         PURPOSE*  
1145           Karachi     2.8  Errand/Supplies  
1146           Karachi     2.9  Errand/Supplies  
1147           Karachi     4.6  Errand/Supplies  
1148           Karachi     4.6          Meeting  
1149           Karachi     0.8   Customer Visit  
1150           Karachi     0.7          Meeting  
1151  Unknown Location     3.9   Temporary Site  
1152  Unknown Location    16.2          Meeting  
1153           Gampaha     6.4   Temporary Site  
1154         Ilukwatta    48.2   Temporary Site

## Q2. Show the first 10 records of the dataset. (2 points)

In [ ]:
dataset.head(10)

START_DATE*         END_DATE* CATEGORY*           START*  \
0  01-01-2016 21:11  01-01-2016 21:17  Business      Fort Pierce   
1  01-02-2016 01:25  01-02-2016 01:37  Business      Fort Pierce   
2  01-02-2016 20:25  01-02-2016 20:38  Business      Fort Pierce   
3  01-05-2016 17:31  01-05-2016 17:45  Business      Fort Pierce   
4  01-06-2016 14:42  01-06-2016 15:49  Business      Fort Pierce   
5  01-06-2016 17:15  01-06-2016 17:19  Business  West Palm Beach   
6  01-06-2016 17:30  01-06-2016 17:35  Business  West Palm Beach   
7  01-07-2016 13:27  01-07-2016 13:33  Business             Cary   
8  01-10-2016 08:05  01-10-2016 08:25  Business             Cary   
9  01-10-2016 12:17  01-10-2016 12:44  Business          Jamaica   

             STOP*  MILES*         PURPOSE*  
0      Fort Pierce     5.1   Meal/Entertain  
1      Fort Pierce     5.0              NaN  
2      Fort Pierce     4.8  Errand/Supplies  
3      Fort Pierce     4.7          Meeting  
4  West Palm Beach    63.7   Customer Visit  
5  West Palm Beach     4.3   Meal/Entertain  
6       Palm Beach     7.1          Meeting  
7             Cary     0.8          Meeting  
8      Morrisville     8.3          Meeting  
9         New York    16.5   Customer Visit

## Q3. Show the dimension(number of rows and columns) of the dataset. (2 points)

In [ ]:
rows = len(dataset.axes[0])
cols = len(dataset.axes[1])
print("Number of Rows: " + str(rows))
print("Number of Columns: " + str(cols))

#Alternatively

dataset.shape

Number of Rows: 1155
Number of Columns: 7


(1155, 7)

## Q4. Show the size (Total number of elements) of the dataset. (2 points)

In [ ]:
dataset.size

8085

## Q5. Display the information about all the variables of the data set. What can you infer from the output?(1 +2 points)

#### Hint: Information includes - Total number of columns,variable data-types, number of non-null values in a variable, and usage

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1155 entries, 0 to 1154
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   START_DATE*  1155 non-null   object 
 1   END_DATE*    1155 non-null   object 
 2   CATEGORY*    1155 non-null   object 
 3   START*       1155 non-null   object 
 4   STOP*        1155 non-null   object 
 5   MILES*       1155 non-null   float64
 6   PURPOSE*     653 non-null    object 
dtypes: float64(1), object(6)
memory usage: 63.3+ KB


The output allows us to infer that there are about 1155 rows in the dataset and 7 columns. The column names are START_DATE\*, END_DATE\*, CATEGORY\*, START\*, STOP\*, MILES\* and PURPOSE\*. Out of these, other than PURPOSE* (653 values), all other columns have 1155 non-null values in total. Other than MILES\* (float64), all others are of object datatype. The datset uses 63.3+ KB in memory.

## Q6. Check for missing values. (2 points) 

####  Note: Output should contain only one boolean value

In [ ]:
dataset.isna().any().any()

True

## Q7. How many missing values are present in the entire dataset? (2 points)


In [ ]:
dataset.isnull().sum().sum()

502

## Q8. Get the summary of the original data. (2 points).

#### Hint: Summary includes- Count,Mean, Std, Min, 25%,50%,75% and max

In [ ]:
dataset.describe()

MILES*
count  1155.000000
mean     10.566840
std      21.579106
min       0.500000
25%       2.900000
50%       6.000000
75%      10.400000
max     310.300000



## Q9. Drop the missing values and store the data in a new dataframe (name it"df") (2-points)

#### Note: Dataframe "df" will not contain any missing value

In [ ]:
df = dataset.dropna()
df.isnull().sum().sum()

0

## Q10. Check the information of the dataframe(df). (1 points)

#### Hint: Information includes - Total number of columns,variable data-types, number of non-null values in a variable, and usage

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 653 entries, 0 to 1154
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   START_DATE*  653 non-null    object 
 1   END_DATE*    653 non-null    object 
 2   CATEGORY*    653 non-null    object 
 3   START*       653 non-null    object 
 4   STOP*        653 non-null    object 
 5   MILES*       653 non-null    float64
 6   PURPOSE*     653 non-null    object 
dtypes: float64(1), object(6)
memory usage: 40.8+ KB


## Q11. Get the unique start locations. (2 points)
#### Note: This question is based on the dataframe with no 'NA' values

In [ ]:
df['START*'].unique()

array(['Fort Pierce', 'West Palm Beach', 'Cary', 'Jamaica', 'New York',
       'Elmhurst', 'Midtown', 'East Harlem', 'Flatiron District',
       'Midtown East', 'Hudson Square', 'Lower Manhattan',
       "Hell's Kitchen", 'Downtown', 'Gulfton', 'Houston', 'Eagan Park',
       'Morrisville', 'Durham', 'Farmington Woods', 'Lake Wellingborough',
       'Fayetteville Street', 'Raleigh', 'Whitebridge', 'Hazelwood',
       'Fairmont', 'Meredith Townes', 'Apex', 'Chapel Hill', 'Northwoods',
       'Edgehill Farms', 'Eastgate', 'East Elmhurst', 'Long Island City',
       'Katunayaka', 'Colombo', 'Nugegoda', 'Unknown Location',
       'Islamabad', 'R?walpindi', 'Noorpur Shahan', 'Preston',
       'Heritage Pines', 'Tanglewood', 'Waverly Place', 'Wayne Ridge',
       'Westpark Place', 'East Austin', 'The Drag', 'South Congress',
       'Georgian Acres', 'North Austin', 'West University', 'Austin',
       'Katy', 'Sharpstown', 'Sugar Land', 'Galveston', 'Port Bolivar',
       'Washington Avenue',

## Q12. What is the total number of unique start destinations? (2 points)
#### Note: Use the original dataframe without dropping 'NA' values

In [ ]:
dataset['START*'].nunique()

176

## Q13. What is the total number of unique stop locations. (2 points)
#### Note: Use the original dataframe without dropping 'NA' values.

In [ ]:
dataset['STOP*'].nunique()

187

## Q14. Display all Uber trips that has the starting point as San Francisco. (2 points)
#### Note: Use the original dataframe without dropping the 'NA' values.


In [ ]:
dataset[dataset['START*'] == 'San Francisco']

START_DATE*         END_DATE* CATEGORY*         START*       STOP*  \
362  05-09-2016 14:39  05-09-2016 15:06  Business  San Francisco   Palo Alto   
440   6/14/2016 16:09   6/14/2016 16:39  Business  San Francisco  Emeryville   
836  10/19/2016 14:02  10/19/2016 14:31  Business  San Francisco    Berkeley   
917  11-07-2016 19:17  11-07-2016 19:57  Business  San Francisco    Berkeley   
919  11-08-2016 12:16  11-08-2016 12:49  Business  San Francisco    Berkeley   
927  11-09-2016 18:40  11-09-2016 19:17  Business  San Francisco     Oakland   
933  11-10-2016 15:17  11-10-2016 15:22  Business  San Francisco     Oakland   
966  11/15/2016 20:44  11/15/2016 21:00  Business  San Francisco    Berkeley   

     MILES*         PURPOSE*  
362    20.5  Between Offices  
440    11.6          Meeting  
836    10.8              NaN  
917    13.2  Between Offices  
919    11.3          Meeting  
927    12.7   Customer Visit  
933     9.9   Temporary Site  
966    11.8   Temporary Site

## Q15. What is the most popular starting point for the Uber drivers? (2 points)
#### Note: Use the original dataframe without dropping the 'NA' values.

#### Hint:Popular means the place that is visited the most

In [ ]:
print(str(dataset['START*'].value_counts().head(1)).split(" ", 1)[0])

Cary


## Q16. What is the most popular dropping point for the Uber drivers? (2 points)
#### Note: Use the original dataframe without dropping the 'NA' values.

#### Hint: Popular means the place that is visited the most

In [ ]:
print(str(dataset['STOP*'].value_counts().head(1)).split(" ", 1)[0])

Cary


## Q17. What is the most frequent route taken by Uber drivers. (3 points)
#### Note: This question is based on the new dataframe with no 'na' values.
#### Hint-Print the most frequent route taken by Uber drivers (Route= combination of START & END points present in the Data set).

In [ ]:
Route = str(pd.DataFrame().assign(COMBINED_START_STOP = df['START*'] + " -> " + df['STOP*'])['COMBINED_START_STOP'].value_counts().head(1)).split(' ',3)[:3]
print(''.join(Route))

Cary->Morrisville


## Q18. Display all types of purposes for the trip in an array. (2 points)
#### Note: This question is based on the new dataframe with no 'NA' values.

In [ ]:
df['PURPOSE*'].nunique()

10

## Q19. Plot a bar graph of Purpose vs Miles(Distance). What can you infer from the plot(2 +2 points)
#### Note: Use the original dataframe without dropping the 'NA' values.
#### Hint:You have to plot total/sum miles per purpose

In [ ]:
dataset_reset = dataset[['PURPOSE*', 'MILES*']].groupby(['PURPOSE*']).sum().reset_index()
dataset_reset.columns = ['Purpose', 'Miles']
fig = px.bar(dataset_reset, x = 'Purpose', y = 'Miles', color = 'Miles')
fig.show()

## Q20. Display a dataframe of Purpose and the total distance travelled for that particular Purpose. (3 points)
#### Note: Use the original dataframe without dropping "NA" values

In [ ]:
dataset[['PURPOSE*','MILES*']].groupby(['PURPOSE*']).sum().sort_values(by = 'MILES*', ascending = False)

MILES*
PURPOSE*               
Meeting          2851.3
Customer Visit   2089.5
Meal/Entertain    911.7
Temporary Site    523.7
Errand/Supplies   508.0
Between Offices   197.0
Commute           180.2
Moving             18.2
Airport/Travel     16.5
Charity ($)        15.1

## Q21. Generate a plot showing count of trips vs category of trips. What can you infer from the plot (2 +1 points)
#### Note: Use the original dataframe without dropping the 'NA' values.

In [ ]:
cat_count = dataset['CATEGORY*'].value_counts()
df_val_counts = pd.DataFrame(cat_count)
df_value_counts_reset = df_val_counts.reset_index()
df_value_counts_reset.columns = ['Category', 'Count Of Trips']
fig = px.bar(df_value_counts_reset, x = 'Count Of Trips', y = 'Category', color = 'Category')
fig.show()

Inference:
Most of the Uber trips are for business purposes as it's count is very high (1078) compared to trips made for personal purposes (only 77). This shows that people prefer Uber mostly for business/official purposes.

## Q22. What percentage of Miles were clocked under Business Category and what percentage of Miles were clocked under Personal Category ? (3 points)

### Note:Use the original dataframe without dropping the 'NA' values. 


In [ ]:
TOTAL = dataset['MILES*'].sum()
MILES = dataset[['CATEGORY*','MILES*']].groupby('CATEGORY*').sum()['MILES*']
BUSINESS = MILES.iloc[:1].sum()
PERSONAL = MILES.iloc[1:].sum()
print('BUSINESS - ' + str((BUSINESS/TOTAL)*100) + ' %\n' + 'PERSONAL - ' + str((PERSONAL/TOTAL)*100) + ' %')

BUSINESS - 94.11947856153776 %
PERSONAL - 5.880521438462232 %


***THE END***